src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/Huffman.cs

```diff
@@ -308,119 +284,152 @@ public static (uint encoded, int bitLength) Encode(int data)
         /// <returns>The number of decoded symbols.</returns>
         public static int Decode(byte[] src, int offset, int count, byte[] dst)
         {
-            int i = offset;
-            int j = 0;
-            int lastDecodedBits = 0;
-            while (i < count)
+            // input validation
+            if (src == null || count == 0)
+                return 0;
+
+            if (dst == null)
+                throw new ArgumentNullException(nameof(dst));
+
+            if ((uint)offset >= (uint)src.Length)
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            if ((uint)count > (uint)(src.Length - offset))
+                throw new ArgumentOutOfRangeException(nameof(count));
+
+            // let's narrow thing down to just the part of the source buffer that we've been asked to decode
+            var sourceSpan = new ReadOnlySpan<byte>(src, offset, count);
+            int sourceIndex = 0;
+            int destinationIndex = 0;
+            int bitOffset = 0;              // symbol bit patterns do not necessarily align to byte boundaries. need to keep track of our offset within a byte
+            var decodingArray = s_decodingArrayLoader.Value;
+            while (sourceIndex < sourceSpan.Length)
             {
-                uint next = (uint)(src[i] << 24 + lastDecodedBits);
-                next |= (i + 1 < src.Length ? (uint)(src[i + 1] << 16 + lastDecodedBits) : 0);
-                next |= (i + 2 < src.Length ? (uint)(src[i + 2] << 8 + lastDecodedBits) : 0);
-                next |= (i + 3 < src.Length ? (uint)(src[i + 3] << lastDecodedBits) : 0);
+                int arrayIndex = 0;         // index into the decoding array
+                int decodedBits = 0;
 
-                uint ones = (uint)(int.MinValue >> (8 - lastDecodedBits - 1));
-                if (i == count - 1 && lastDecodedBits > 0 && (next & ones) == ones)
+                // decoding a symbol will take a maximum of four bytes (longest symbol is represented by a 30 bit pattern)
+                for (int i = 0; i <= 24; i += 8)
                 {
-                    // The remaining 7 or less bits are all 1, which is padding.
-                    // We specifically check that lastDecodedBits > 0 because padding
-                    // longer than 7 bits should be treated as a decoding error.
-                    // http://httpwg.org/specs/rfc7541.html#rfc.section.5.2
-                    break;
-                }
+                    // grab the next byte and push it over to make room for bits we have to borrow
+                    byte workingByte = (byte)(sourceSpan[sourceIndex] << bitOffset);
+                    // borrow some bits if we need to
+                    if (bitOffset > 0)
+                        workingByte |= (byte)((sourceIndex < sourceSpan.Length - 1 ? sourceSpan[sourceIndex + 1] : 0) >> (8 - bitOffset));
```

Comment:
Consider expressing this as a nested 'if` - it's hard to eyeball what's going on here, and the `false` branch does nothing

Reply:
Changed.